#97 Implemented delayed validation by introducing delayedValidation o…

andkorsh opened the pull request:
#97 Implemented delayed validation by introducing delayedValidation option.

The logic is a little bit more complicated than switching validation event to blur as initially discussed in the issue. This implementation doesn't do any validation while the user typing for the first time (i.e. until the first 'change' or 'submit') but once the validation had failed and user focused back into the input, validation switches back to "input" event to remove the error as soon as the data is valid.

haacked commented on src/index.ts:
```diff
@@ -1068,6 +1069,15 @@ export class ValidationService {
+                !input.classList.contains(this.ValidationInputCssClassName)
+            ) {
+                // When delayedValidation=true, "input" only takes it back to valid. "Change" can make it invalid.
+                return;
```
This needs to return `true` or `false`.

haacked commented:
Do you mind adding a sample page with delayed validation enabled so it's easy to test the behavior?

andkorsh commented:
@haacked sure, thanks for looking at this. Fixed the return statement and added a sample. Please have a look when the time allows. We'd love to see it in the package to use it in our app.

Also below is a gif demonstrating the behavior of new delayed validation. Idea is, it doesn't bother you while you still typing, but only shows up when you finished and focused out/submitted. And if the field is errored and you are back in it to fix it, it will turn back valid as soon as the data becomes valid.

![delayed_validation](https://github.com/user-attachments/assets/4ea926d0-308a-4572-b12c-76e00283e37b)

In fact, the original lib at https://jqueryvalidation.org/files/demo/ behaves very similarly:

![delayed_validation (original)](https://github.com/user-attachments/assets/164cc68d-f8b0-4c2e-9311-715debab7c84)

haacked commented:
Oh yeah, the current behavior is annoying. It seems to me the behavior in this PR should be the default as it seems to align with the jquery validation behavior. @dahlbyk any thoughts?

andkorsh commented:
@haacked If we choose to make this behavior default, I think it would be better to remove the `delayedValidation` option altogether, leaving users with `data-val-event="input"` should they require the old behavior. Also we'd need to refactor the below lines and get rid of nested ternary. When I implemented this PR, I tried not to touch any of the existing behaviors and did not fix the bug where `data-val-event` value isn't considered at all. But if we do choose to introduce a change in default behavior, fixing the bug would be logical.

https://github.com/haacked/aspnet-client-validation/blob/a59c47aebbc94cb3e930c3685c6865c8963a183a/src/index.ts#L1102-L1104

 Please let me know what you decide and I can take care of it.

cheers,
Andrew

haacked commented:
@andkorsh I agree with your plan. Let’s make it so.

andkorsh commented:
@haacked I have updated the PR now and created a "Immediate Validation" sample page that comparing the new default behavior with the old behavior which can now be achieved using `data-val-event="input"`.

cheers!

haacked commented on src/index.ts:
```diff
@@ -1068,6 +1068,15 @@ export class ValidationService {
             let validate = this.validators[uid];
             if (!validate) return true;
+            
+            if (
+                !input.dataset.valEvent &&
+                event && event.type === 'input' &&
+                !input.classList.contains(this.ValidationInputCssClassName)
+            ) {
```
Minor nit, but I prefer the logical operators in a multi-line conditional to align on the left. Makes it easier to see all the component conditionals.

```suggestion
            if (!input.dataset.valEvent
                && event
                && event.type === 'input'
                && !input.classList.contains(this.ValidationInputCssClassName)
            ) {
```

haacked reviewed (commented):
Looks good! I need to test it before I approve it.

andkorsh commented:
Thank you! And appreciate you creating this great package, hope Microsoft will use it and remove the jquery dependency from identity templates one day.

andkorsh commented:
It is still a drop in replacement and behavior now aligns more with how the
original library behaves.

cheers

On Sat, Aug 17, 2024 at 3:59 AM lonix1 ***@***.***> wrote:

> @andkorsh <https://github.com/andkorsh> As discussed in original issue,
> is the final result still a drop-in replacement? If not, what should I
> change in my code to upgrade to these bits?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/haacked/aspnet-client-validation/pull/119#issuecomment-2294799821>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA5YBM6TIGOX5UAWNAVBEODZR4GGFAVCNFSM6AAAAABMHMCRF2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEOJUG44TSOBSGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
